src/mlpack/core/dists/gamma_distribution.cpp

```diff
@@ -53,6 +64,13 @@ void GammaDistribution::Train(const arma::mat& rdata, const double tol)
   Train(logMeanxVec, meanLogxVec, meanxVec, tol);
 }
 
+// Fits an alpha and beta parameter according to observation probabilities.
+void GammaDistribution::Train(const arma::mat& observations, 
+                              const arma::vec& probabilities,
+                              const double tol)
+{
```

Comment:
Hm, maybe that will work; I haven't thought about it too much.  I think the only issue might be that if you take unlikely points to have low probability, then when you train, this biases the training points towards the high-PDF parts of the distribution, possibly giving (I think) a trained distribution with different properties than the original.  If it was a one-dimensional Gaussian we were training, this would result in a lower variance, but I don't have the intuition to say what it will do the Gamma distribution, only that I think you'll end up with a different Gamma distribution than the one you are taking random samples from.  Hopefully what I've written here is at least somewhat coherent. :)